File: src/types/metrics.ts
/**
 * TypeScript interfaces for metrics data from Rust backend
 * Matches the structures in og-types crate
 */

export interface RawMetrics {
  pagerankImports: number;
  pagerankCalls?: number;
  indegree: number;
  outdegree: number;
  kCore: number;
  clustering: number;
  betweenness: number;
  churn: number;
  complexity: number;
  owners: number;
  coverage: number;
}

export interface NormalizedMetrics {
  pagerankImports: number;
  pagerankCalls?: number;
  indegree: number;
  kCore: number;
  clustering: number;
  betweenness: number;
  churn: number;
  complexity: number;
  owners: number;
  coverage: number;
}

export interface UINodeMetricsV1 {
  path: string;
  name: string;
  nodeType: string;
  community: number;
  importance: number;
  risk: number;
  chokepoint: number;
  payoff: number;
  raw: RawMetrics;
  normalized: NormalizedMetrics;
  version?: number;
}

export interface AnalysisSummary {
  totalNodes: number;
  totalEdges: number;
  numCommunities: number;
  modularity: number;
  avgComplexity: number;
  highRiskCount: number;
  circularDependencies: number;
}

export interface AnalyzedGraph {
  graphData: GraphData;
  metrics: UINodeMetricsV1[];
  summary: AnalysisSummary;
}

export interface GraphNode {
  id: string;
  name: string;
  type: string;
  filePath?: string;
  nodeType?: string;
  fileType?: string;
  // Add metrics reference
  metrics?: UINodeMetricsV1;
}

export interface GraphLink {
  source: string;
  target: string;
  type?: string;
  linkType?: string;
  value?: number;
  strength?: number;
}

export interface GraphData {
  nodes: GraphNode[];
  links: GraphLink[];
  stats?: {
    fileCount: number;
    nodeCount: number;
    linkCount: number;
  };
}

// Helper function to find metrics for a node
export function findMetricsForNode(
  nodeId: string,
  metrics: UINodeMetricsV1[]
): UINodeMetricsV1 | undefined {
  // Node ID might be the path or a variation of it
  return metrics.find(m => 
    m.path === nodeId || 
    m.path.endsWith(nodeId) ||
    nodeId.endsWith(m.path)
  );
}

// Color palette for communities
export const COMMUNITY_COLORS = [
  '#3498db', // Blue
  '#e74c3c', // Red
  '#2ecc71', // Green
  '#f39c12', // Orange
  '#9b59b6', // Purple
  '#1abc9c', // Turquoise
  '#34495e', // Dark Gray
  '#f1c40f', // Yellow
  '#e67e22', // Carrot
  '#95a5a6', // Light Gray
];

export function getCommunityColor(communityId: number): string {
  return COMMUNITY_COLORS[communityId % COMMUNITY_COLORS.length];
}

// Scale functions for visual encoding
export function scaleNodeSize(importance: number): number {
  // Scale from 0.5x to 2x based on importance (0-1)
  return 0.5 + importance * 1.5;
}

export function shouldShowRiskIndicator(risk: number): boolean {
  return risk > 0.7;
}

export function shouldShowChokepointGlow(chokepoint: number): boolean {
  return chokepoint > 0.7;
}

File: crates/og-types/src/metrics.rs
use serde::{Deserialize, Serialize};

/// File-level metrics
#[derive(Debug, Clone, Default)]
pub struct FileMetrics {
    pub lines_of_code: usize,
    pub complexity: usize,
    pub functions: usize,
    pub classes: usize,
    pub imports: usize,
    pub exports: usize,
}

/// UI-facing stable contract for node metrics
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct UINodeMetricsV1 {
    pub path: String,
    pub name: String,
    pub node_type: String,
    pub community: i64,
    pub importance: f32,
    pub risk: f32,
    pub chokepoint: f32,
    pub payoff: f32,
    pub raw: RawMetrics,
    pub normalized: NormalizedMetrics,
    #[serde(default = "default_version")]
    pub version: u8,
}

fn default_version() -> u8 {
    1
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RawMetrics {
    pub pagerank_imports: f64,
    pub pagerank_calls: Option<f64>,
    pub indegree: i64,
    pub outdegree: i64,
    pub k_core: i64,
    pub clustering: f64,
    pub betweenness: f64,
    pub churn: i64,
    pub complexity: i64,
    pub owners: i64,
    pub coverage: f64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct NormalizedMetrics {
    pub pagerank_imports: f64,
    pub pagerank_calls: Option<f64>,
    pub indegree: f64,
    pub k_core: f64,
    pub clustering: f64,
    pub betweenness: f64,
    pub churn: f64,
    pub complexity: f64,
    pub owners: f64,
    pub coverage: f64,
}

/// Composite metric outputs
#[derive(Debug, Clone)]
pub struct CompositeOutputs {
    pub importance: f64,
    pub chokepoint: f64,
    pub risk: f64,
    pub payoff: f64,
}

/// Normalization ranges for metrics
#[derive(Debug, Clone)]
pub struct NormalizationRanges {
    pub pagerank_imports: (f64, f64),
    pub pagerank_calls: (f64, f64),
    pub k_core: (f64, f64),
    pub indegree: (f64, f64),
    pub clustering: (f64, f64),
    pub betweenness: (f64, f64),
    pub churn: (f64, f64),
    pub complexity: (f64, f64),
    pub owners: (f64, f64),
    pub coverage: (f64, f64),
}